miden-crypto/src/merkle/smt/historical/mod.rs

```diff
@@ -0,0 +1,500 @@
+        kv_pairs: impl IntoIterator<Item = (Word, Word)>,
+    ) -> Result<MutationSet<SMT_DEPTH, Word, Word>, HistoricalError> {
+        let inner = self.inner.read().unwrap();
+        Ok(inner.latest.compute_mutations(kv_pairs)?)
+    }
+
+    /// Apply the given forward mutation set to the interior [`Smt`].
+    /// This method uses interior mutability with RwLock, ensuring only one thread can write at a
+    /// time.
+    ///
+    /// Creates a reversion `MutationSet` to be able to reconstruct the previous state of the `Smt`
+    /// on-demand and applies the forward mutations as expected using [`Smt::apply_mutations`].
+    pub fn apply_mutations(
+        &self,
+        mutation_set: MutationSet<SMT_DEPTH, Word, Word>,
+    ) -> Result<(), HistoricalError> {
+        let mut inner = self.inner.write().unwrap();
+
+        // Apply forward mutations directly to the Smt (we only have one in memory)
+        let reversion = inner
+            .latest
+            .apply_mutations_with_reversion(mutation_set)
+            .map_err(HistoricalError::MerkleError)?;
+
+        // Track the historical state with Arc for efficient sharing
+        let state = HistoricalReversion::from_reversion_mutation_set(&mut inner, reversion);
+        inner.history.push_front(Arc::new(state));
+        Self::cleanup(&mut inner.history);
+
+        Ok(())
+    }
+
+    #[cfg(test)]
+    pub(crate) fn first_reversion(&self) -> Option<MutationSet<SMT_DEPTH, Word, Word>> {
+        let inner = self.inner.read().unwrap();
+        inner.history.front().map(|state| (*state.mutations).clone())
+    }
+}
+
+/// A historical view of the `Smt`
+///
+/// Attention: Holds a `RwLockReadGuard` to ensure the underlying `Smt` from being modified while
+/// the view exists.
+pub struct HistoricalView<'a> {
+    /// Hold onto the `latest: Smt` to ensure it doesn't change while we use the overlay.
+    inner: RwLockReadGuard<'a, InnerState>,
+    /// The reversions needed to reach this historical state from `latest`.
+    ///
+    /// `latest + [0, 1, 2, 3, 4, 5... MAX_HISTORY] -> state at[n-1] steps into the past`
+    /// where `n` is implicitly encoded in the `reversions.len()`.
+    reversions: Vec<Arc<HistoricalReversion>>,
+}
+
+impl<'a> HistoricalView<'a> {
+    /// An iterator for the reversion stack
+    ///
+    /// Flattens all mutations from all historical states, to be applied in order.
+    /// The order is from newest to oldest.
+    fn reversion_iter(
+        &self,
+    ) -> impl DoubleEndedIterator<Item = &MutationSet<SMT_DEPTH, Word, Word>> + '_ {
+        self.reversions.iter().map(|state| state.mutations.as_ref())
+    }
+
+    /// Root of the historical view
+    pub fn root(&self) -> Word {
+        // If we have reversions, the last one contains the historical root we want
+        self.reversions
+            .last()
+            .map(|state| state.root)
+            .unwrap_or_else(|| self.inner.latest.root())
+    }
+
+    /// Wrapper.
```

Comment:
The get_node_hash method is a natural place to implement fine-grained locking. Replace current reversion lookup with a DashMap<NodeIndex, parking_lot::RwLock<Word>> cache. This would allow concurrent access to different tree branches while preparing the foundation for future Rayon parallel hashing implementation.